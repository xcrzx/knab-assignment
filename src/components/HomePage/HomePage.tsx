import {
  Button,
  Grid,
  makeStyles,
  TextField,
  Typography,
} from "@material-ui/core";
import axios from "axios";
import { PageLayout } from "components/PageLayout/PageLayout";
import React, { useCallback } from "react";

const useStyles = makeStyles(() => ({
  form: {
    display: "flex",
    alignItems: "baseline",
  },
}));

export const HomePage = () => {
  const classes = useStyles();
  const [baseCurrency, setBaseCurrency] = React.useState("");
  const [exchangeRates, setExchangeRates] = React.useState<
    [currency: string, rate: number][]
  >();
  const [error, setError] = React.useState();

  const handleBaseCurrencyChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setExchangeRates(undefined);
      setBaseCurrency(e.target.value);
    },
    []
  );

  const handleGetQuotes = useCallback(async () => {
    setError(undefined);
    setExchangeRates(undefined);
    try {
      const { data } = await axios.get<[currency: string, rate: number][]>(
        "/api/exchange-rates",
        {
          params: { baseCurrency },
        }
      );

      setExchangeRates(data);
    } catch (e) {
      setError(e.message);
    }
  }, [baseCurrency]);

  return (
    <PageLayout title="GetQuote">
      <Grid container spacing={4}>
        <Grid item xs={12} className={classes.form}>
          <TextField
            label="Base Currency"
            value={baseCurrency}
            onChange={handleBaseCurrencyChange}
          />
          <Button onClick={handleGetQuotes}>Get Quotes</Button>
        </Grid>
        {error && (
          <Grid item>
            <Typography variant="body1" color="error">
              {error}
            </Typography>
          </Grid>
        )}
        {exchangeRates && (
          <Grid item component="output">
            <Typography variant="body1">1 {baseCurrency} =</Typography>
            {exchangeRates.map(([currency, rate]) => (
              <Typography variant="body1" key={currency}>
                {new Intl.NumberFormat("en-US", {
                  style: "currency",
                  currency,
                }).format(rate)}
              </Typography>
            ))}
          </Grid>
        )}
      </Grid>
    </PageLayout>
  );
};
