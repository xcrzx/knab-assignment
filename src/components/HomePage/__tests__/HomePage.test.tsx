import { HomePage } from "../HomePage";
import { render } from "@testing-library/react";

describe("HomePage Component", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
