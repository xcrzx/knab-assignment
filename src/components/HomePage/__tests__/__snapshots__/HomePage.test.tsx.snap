// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-layout-2"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          GetQuote
        </h6>
      </div>
    </header>
    <main
      class="MuiContainer-root makeStyles-content-3 MuiContainer-maxWidthLg"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
      >
        <div
          class="MuiGrid-root makeStyles-form-1 MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Base Currency
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Get Quotes
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </main>
    <footer
      class="MuiContainer-root makeStyles-footer-4 MuiContainer-maxWidthLg"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        Copyright Â© 2021 Get Quote Inc. All Rights Reserved
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
